### DISCUSSION TITLE
Add `$this` to attribute callback to access other model properties

### DISCUSSION THREAD

**Discussion by ben221199** [laravel/framework#52208] [Ideas] [answered]
https://github.com/laravel/framework/blob/9d7c869cd53c5c18b1939a7d826506f31f74669c/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L706

In the line above, only `$value` and `$this->attributes` are passed to the closure. However, I think it would be useful to add `$this` as third parameter, so that I also can access other model properties, like relations. Now I only can access attributes.

**Comment by ghost** [comment]
Hi. A trait is not aware where it is being used so your desire is not recommended.

**Reply by ben221199** [reply]
What would you advise me to do then?

**Reply by ghost** [reply]
Set the relations using setAttribute if you really need them like that.

**Reply by ghost** [reply]
Or you can try to use $this in that callback because it should work. You may test it out.

**Reply by ben221199** [reply]
If I remember correctly, that failed. I will test it again today.

**Reply by ben221199** [reply]
`Using $this when not in object context`

**Reply by ben221199** [reply]
The `get_defined_vars()` only gives `$value` and `$attributes`.

**Reply by ghost** [reply]
The setting the relation you need as attribute is the only way that I can think of.

**Comment by henzeb** [comment]
you can of course bind your Closure to $this when passing it to the Attribute?

```php
Attribute::get(
    Closure::bind(
        function(){}, 
        $this
    )
);
```

or 

```php
Attribute::get(
    (function(){})->bindTo($this)
);
```

**Reply by ben221199** [reply]
Is that the same as `$record = $this; static function(mixed $value,array $attributes) use($record){ /*code*/ }`?

**Reply by ben221199** [reply]
Never mind.

**Comment by ben221199** [comment, marked as answer]
Okay, I know the problem. This was my code I hoped would work:

```php
protected function nameFinal(): Attribute{
	return Attribute::make(static function(mixed $value,array $attributes){
		dump(get_defined_vars(),$this);
	})->shouldCache();
}
```

But this didn't work, so I made it like this:

```php
protected function nameFinal(): Attribute{
	$record = $this;
	return Attribute::make(static function(mixed $value,array $attributes) use($record){
		dump(get_defined_vars(),$record);
	})->shouldCache();
}
```

However, it can be much simpler. First of all, `get_defined_vars()` will not list `$this`, so only `$value` and `$attributes` are visible. It isn't a good function to see if you have access to `$this`. The real problem was that I used `static function`, which doesn't allow `$this`. Removing the `static` keyword makes it possible to use `$this`. I don't even have to use `bindTo()` on the closure.

The new code looks like this:
```php
protected function nameFinal(): Attribute{
	return Attribute::make(function(mixed $value,array $attributes){
		dump($this);
	})->shouldCache();
}
```

**Comment by ghost** [comment]
@ben221199 Thanks for asking this question in an incomplete manner :) And I'm not joking. I learned new things from this thread.

**Reply by ben221199** [reply]
Do you mean I did a pore job on asking a question? 😂

**Reply by ghost** [reply]
I was unaware that you can declare a callback static for example. Also that you can manually bind it. I knew from macros that this is available in the callback.
